File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
    if (head === null || k <= 1) return head

    const dummy = new ListNode(0)
    dummy.next = head
    let groupPrev = dummy

    while (true) {
        let kth = groupPrev
        for (let i = 0; i < k && kth !== null; i++) {
            kth = kth.next
        }

        if (kth === null) {
            break
        }

        let groupNext = kth.next
        let prev = groupNext
        let current = groupPrev.next

        while (current !== groupNext) {
            let temp = current.next
            current.next = prev
            prev = current
            current = temp
        }

        let temp = groupPrev.next
        groupPrev.next = kth
        groupPrev = temp
    }

    return dummy.next
};

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        def get_length(node):
            length = 0
            
            while node:
                node = node.next
                length += 1
            
            return length

        length = get_length(head)
        dummy = ListNode(0, head)
        prev_group = dummy

        while length >= k:
            prev = None
            curr = prev_group.next
            tail = curr

            for _ in range(k):
                nxt = curr.next
                curr.next = prev
                prev = curr
                curr = nxt

            prev_group.next = prev
            tail.next = curr
            prev_group = tail

            length -= k
        
        return dummy.next